scipy/stats/_morestats.py

```diff
@@ -3293,78 +3293,69 @@ def func(x):
-    def psi(indicator):
-        return (indicator - (N + 1)/2)**2
-
-    # define summation range for use in calculation of phi, as seen in sum
-    # in the unnumbered equation on the bottom of page 312 (Mielke).
-    s_lower = S[js - 1] + 1
-    s_upper = S[js] + 1
-    phi_J = [np.arange(s_lower[idx], s_upper[idx]) for idx in range(k)]
-
-    # for every range in the above array, determine the sum of psi(I) for
-    # every element in the range. Divide all the sums by `t`. Following the
-    # last unnumbered equation on page 312.
-    phis = [np.sum(psi(I_j)) for I_j in phi_J] / t[js]
-
-    # `T` is equal to a[j] * phi[j], per the first unnumbered equation on
-    # page 312. `phis` is already in the order based on `js`, so we index
-    # into `a` with `js` as well.
-    T = sum(phis * a[js])
-
-    # The approximate statistic
+def _mood_statistic_with_ties(x, y, t, n, m, N):
+    # First equation of "Mood's Squared Rank Test", Mielke pg 313
     E_0_T = n * (N * N - 1) / 12
 
-    varM = (m * n * (N + 1.0) * (N ** 2 - 4) / 180 -
-            m * n / (180 * N * (N - 1)) * np.sum(
-                t * (t**2 - 1) * (t**2 - 4 + (15 * (N - S - S_i_m1) ** 2))
-            ))
-
-    return ((T - E_0_T) / np.sqrt(varM),)
+    # m, n, N, t, and S are defined in the second paragraph of Mielke pg 312
+    # The only difference is that our `t` has zeros interspersed with the relevant
+    # numbers to keep the array rectangular, but these terms nothing to the sum.
+    np = array_namespace(x, y, t)  # needed to use `cumulative_sum`
+    S = np.cumulative_sum(t, include_initial=True, axis=-1)
+    S_i, S_i_m1 = S[..., 1:], S[..., :-1]
+    # Second equation of "Mood's Squared Rank Test", Mielke pg 313
+    varM = (m * n * (N + 1.0) * (N**2 - 4) / 180
+            - m * n / (180 * N * (N - 1))
+            * np.sum(t * (t ** 2 - 1) * (t ** 2 - 4 + (15 * (N - S_i - S_i_m1) ** 2)),
+                     axis=-1))
+
+    # There is a formula for phi in terms of t, S, and Psi(I) = [I - (N+1)/2]^2 is
+    # defined (with a mistake in the location of the ^2) at the beginning of
+    # "Mood's Squared Rank Test", Mielke pg 313. To vectorize this calculation,
+    # let c = (N + 1) / 2, so Psi(I) = I^2 - 2*c*I + c^2. We sum each of these three
+    # parts of Psi separately using formulas for sums from a to b (inclusive) of
+    # terms I^2, I, and 1 where I takes on successive integers.
+    def sum_I2(a, b):
+        return (b * (b + 1) * (2*b + 1) - a * (a - 1) * (2*a - 1)) / 6
+
+    def sum_I(a, b):
+        return (b * (b + 1) - a * (a - 1)) / 2
+
+    def sum_1(a, b):
+        return (b - a) + 1
+
+    with np.errstate(invalid='ignore', divide='ignore'):
+        sum_I2 = sum_I2(S_i_m1 + 1, S_i)
+        sum_I = sum_I(S_i_m1 + 1, S_i)
+        sum_1 = sum_1(S_i_m1 + 1, S_i)
+        c = (N + 1) / 2
+        phi = (sum_I2 - 2*c*sum_I + sum_1*c**2) / t
+
+    phi[t == 0] = 0  # where t = 0 we get NaNs; eliminate them
```

Comment:
It is not SO bad to compare the implementation of the code above to the math if you assume take `Si` to represent $S_i$ and `S_i_m1` to represent $S_{i-1}$ at face value. The tricky thing is those zeros in `t` and corresponding duplicate elements in `S_i` and `S_i_m1`. But it all works out:
- The unique terms we care about are calculated because `t`, `S_i`, and `S_i_m1` are all aligned as we expect.
- The terms we don't care about turn into NaNs because at these locations, `t == 0` and `S_i == S_i_m1`.
- We replace the NaNs with zeros so they don't bite us later.